src/matrix/kaldi-matrix.cc

```diff
@@ -2784,6 +2784,78 @@ void MatrixBase<Real>::DiffTanh(const MatrixBase<Real> &value,
 
 
 template<typename Real>
+void MatrixBase<Real>::AddSpatialRegularizationDeriv(
+    const MatrixBase<Real>& out_value, Real scale, Real* regularization_sumsq) {
+  const unsigned int kBlockSize = 256;
+  const unsigned int kSectWidth = 16;
+  KALDI_ASSERT(SameDim(*this, out_value));
+
+  int ext_cols = ((this->NumCols() + (kSectWidth - 1)) / kSectWidth)
+      * kSectWidth;
+  Matrix<Real> val_ext(this->NumRows(), ext_cols);
+  Matrix<Real> conv1(this->NumRows(), ext_cols);
+  Matrix<Real> conv2(this->NumRows(), ext_cols);
+
+  // add 0 padding to the end of the rows
+  for (int32 i = 0; i < this->NumRows(); i++) {
+    for (int32 j = 0; j < this->NumCols(); j++) {
+      val_ext(i, j) = out_value(i, j);
+    }
+  }
+  const int image_rows = ext_cols / kSectWidth;
+  for (int32 i = 0; i < this->NumRows(); i++) {
+    for (int32 j = 0; j < ext_cols; j++) {
+      int image_x = j % kSectWidth;
+      int image_x_left = (image_x + (kSectWidth - 1)) % kSectWidth;
+      int image_x_right = (image_x + 1) % kSectWidth;
+      int image_y = j / kSectWidth;
+      int image_y_up = (image_y + (image_rows - 1)) % image_rows;
+      int image_y_down = (image_y + 1) % image_rows;
+      conv1(i, j) = val_ext(i, j)
+          - 0.125
+              * (val_ext(i, image_y_up * kSectWidth + image_x_left)
+                  + val_ext(i, image_y_up * kSectWidth + image_x)
+                  + val_ext(i, image_y_up * kSectWidth + image_x_right)
+                  + val_ext(i, image_y * kSectWidth + image_x_left)
+                  + val_ext(i, image_y * kSectWidth + image_x_right)
+                  + val_ext(i, image_y_down * kSectWidth + image_x_left)
+                  + val_ext(i, image_y_down * kSectWidth + image_x)
+                  + val_ext(i, image_y_down * kSectWidth + image_x_right));
+    }
+  }
+  if (regularization_sumsq) {
+    Real sum_squared = Real(0);
+    for (int32 i = 0; i < this->NumRows(); i++) {
+      for (int32 j = 0; j < this->NumCols(); j++) {
+        sum_squared += conv1(i, j) * conv1(i, j);
+      }
+    }
+    *regularization_sumsq = sum_squared;
+  }
+  for (int32 i = 0; i < this->NumRows(); i++) {
```

Comment:
best to use a variable for this->NumRows() and this->NumCols() when used as a loop test... avoids function call.